### PR TITLE
[BEAM-10641] Add eliminate_common_key_with_none graph optimizer

### PR DESCRIPTION
**Pull request by yifanmai**
If multiple `KeyWithNone` stages share a common input, then all but one stages will be eliminated along with their output `PCollection`s. Transforms that originally read input from the output `PCollection` of the eliminated `KeyWithNone` stages will be remapped to read input from the output `PCollection` of the remaining `KeyWithNone` stage.

This allows `pack_combiners` to pack sibling `CombineGlobally` stages, rather than just `CombinePerKey` stages.

------------------------
Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Dataflow | Flink | Samza | Spark | Twister2
--- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/) | ---
Java | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Java11/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Java11/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Twister2/lastCompletedBuild/)
Python | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python38/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_VR_Dataflow_V2/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/) | ---
XLang | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Direct/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | [Build Status](https://ci-beam.apache.org/job/beam_PostCommit_XVR_Spark/lastCompletedBuild/) | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website | Whitespace | Typescript
--- | --- | --- | --- | --- | --- | ---
Non-portable | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/)<br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocker_Cron/lastCompletedBuild/) <br>[Build Status](https://ci-beam.apache.org/job/beam_PreCommit_PythonDocs_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Whitespace_Cron/lastCompletedBuild/) | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Typescript_Cron/lastCompletedBuild/)
Portable | --- | [Build Status](https://ci-beam.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | --- | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by yifanmai** [comment]
R: @robertwb

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 1]
```diff
@@ -704,6 +705,64 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
...
+  read input from the output of the eliminated KeyWithNone stages will be
+  remapped to read input from the output of the remaining KeyWithNone stage.
+  """
+  # Partition stages by whether they are eligible for common KeyWithNone
+  # elimination, and group eligible KeyWithNone stages by parent and
+  # environment.
+  grouped_eligible_stages = collections.defaultdict(list)
+  ineligible_stages = []
+  for stage in stages:
+    is_eligible = False
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:724]: Most of this is very general. What if you had a `merge_key(stage)` helper function which only returned something for this case (but was generally extensible) and then threw stuff into the dict according to this key (say, if it was not None), and then the rest of the logic could follow.

**Review by robertwb — COMMENTED** [review]
Looks solid, just some minor suggestions.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 1 above)
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: Refactored the code to separate the key function from the grouping function, and used the grouping function in combiner packing as well; PTAL.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 2]
```diff
@@ -704,6 +705,64 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
...
+  pcoll_id_remap = {}
+  for sibling_stages in grouped_eligible_stages.values():
+    output_pcoll_ids = [
+        only_element(stage.transforms[0].outputs.values())
+        for stage in sibling_stages
+    ]
+    for to_delete_pcoll_id in output_pcoll_ids[1:]:
+      pcoll_id_remap[to_delete_pcoll_id] = output_pcoll_ids[0]
+      del context.components.pcollections[to_delete_pcoll_id]
+    del sibling_stages[1:]
```
[the remark this answers is not in the document]
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: We need to use the loop below to remap the PCollections so I made a new flat list.

**Comment by yifanmai** [comment]
I'm confused by the yapf failure: when I run yapf locally, it produces the exact opposite recommendation as the precommit.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 3]
```diff
@@ -704,6 +705,71 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
...
+    for to_delete_pcoll_id in output_pcoll_ids[1:]:
+      pcoll_id_remap[to_delete_pcoll_id] = output_pcoll_ids[0]
+      del context.components.pcollections[to_delete_pcoll_id]
+    remaining_stages.append(sibling_stages[0])
+
+  # Yield stages while remapping input PCollections if needed.
+  stages_to_yield = itertools.chain(ineligible_stages, remaining_stages)
+  for stage in stages_to_yield:
+    for transform in stage.transforms:
+      for input_key in list(transform.inputs):
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:766]: Use `transform.inputs.keys()`.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py [hunk 4]
```diff
@@ -704,6 +705,71 @@ def fix_side_input_pcoll_coders(stages, pipeline_context):
...
+        pardo_payload = proto_utils.parse_Bytes(
+            transform.spec.payload, beam_runner_api_pb2.ParDoPayload)
+        if pardo_payload.do_fn.urn == python_urns.KEY_WITH_NONE_DOFN:
+          return (only_element(transform.inputs.values()), stage.environment)
+    return None
+
+  grouped_eligible_stages, ineligible_stages = _group_stages_by_key(
+      stages, get_stage_key)
+
+  # Eliminate stages and build the PCollection remapping dictionary.
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:749]: Just a thought: most of this is generic too (parameterized by a `merge_stages : stages -> stage(s), pcoll_remap`).

**Review by robertwb — APPROVED** [review]
Not sure about yapf, perhaps you're using a different version? If you can't get it to work, just manually patch it up to match what jenkins wants.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 3 above)
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py (hunk 4 above)
Comment (yifanmai) [reply sdks/python/apache_beam/runners/portability/fn_api_runner/translations.py:749]: Yeah this is basically a combine by key :) but I think I will leave it as is to keep it readable.

**Comment by yifanmai** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by yifanmai** [comment]
OK, this should be ready for merge.

**Pull request merged**